### PR TITLE
Added client side reciept verification; Implemented subscription upgrade/downgrade

### PR DESCRIPTION
**Pull request by zkzhao333**

### CODE REVIEW

File: GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Controller/PurchaseController.cs [hunk 1]
```diff
@@ -158,10 +159,23 @@ public void OnInitializeFailed(InitializationFailureReason error)
...
 
+    private void InitSubscriptionsBasedOnReceipt(IStoreController controller)
+    {
+        var silverSubscriptionProduct = controller.products.WithID(SubscriptionList.SilverSubscription.ProductId);
+        var goldenSubscriptionProduct = controller.products.WithID(SubscriptionList.GoldenSubscription.ProductId);
+        if (!(silverSubscriptionProduct.hasReceipt && ReceiptValidation(silverSubscriptionProduct.receipt)) &&
+            !(goldenSubscriptionProduct.hasReceipt && ReceiptValidation(goldenSubscriptionProduct.receipt)))
+        {
+            GameDataController.GetGameData().Unsubscribe();
+            Debug.Log("No subscription receipt found. Unsubscribe all subscriptions");
```
Comment (Qing451800) [inline GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Controller/PurchaseController.cs]: What are you doing here? Why do you unsubscribe?

File: GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Controller/PurchaseController.cs [hunk 2]
```diff
@@ -158,10 +159,23 @@ public void OnInitializeFailed(InitializationFailureReason error)
...
+        }
+    }
+
     private static void DeferredPurchase(string productId)
```
Comment (Qing451800) [inline GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Controller/PurchaseController.cs]: ProcessDeferredPurchase

File: GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Controller/PurchaseController.cs [hunk 3]
```diff
@@ -174,19 +188,43 @@ private static void DeferredPurchase(string productId)
...
+        Debug.LogError("Product ID doesn't match any of exist one-time products that can be deferred purchase.");
     }
 
     public PurchaseProcessingResult ProcessPurchase(PurchaseEventArgs args)
     {
         // TODO: Use an error dialog to communicate.
         Debug.Log($"ProcessPurchase: PASS. Product: '{args.purchasedProduct.definition.id}'");
+        bool validPurchase = ReceiptValidation(args.purchasedProduct.receipt);
+
+        if (validPurchase)
```
Comment (Qing451800) [inline GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Controller/PurchaseController.cs]: if (ReceiptValidation(args.purchasedProduct.receipt))

File: GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Controller/PurchaseController.cs [hunk 4]
```diff
@@ -174,19 +188,43 @@ private static void DeferredPurchase(string productId)
...
+
+        // Return a flag indicating whether this product has completely been received, or if the application needs 
+        // to be reminded of this purchase at next app launch. Use PurchaseProcessingResult.Pending when still 
+        // saving purchased products to the cloud, and when that save is delayed. 
+        return PurchaseProcessingResult.Complete;
+    }
+
+    private static bool ReceiptValidation(string receipt)
+    {
+        bool validPurchase = true;
```
Comment (Qing451800) [inline GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Controller/PurchaseController.cs]: Do you need this line?

File: GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Controller/PurchaseController.cs [hunk 5]
```diff
@@ -174,19 +188,43 @@ private static void DeferredPurchase(string productId)
...
+        // Do client side receipt validation.
+        validPurchase = ClientSideReceiptValidation(receipt);
+#endif
+        return validPurchase;
+    }
+
+    private static bool ClientSideReceiptValidation(string receipt)
+    {
+        bool validPurchase = true;
 #if UNITY_ANDROID
```
Comment (Qing451800) [inline GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Controller/PurchaseController.cs:216]: What does this check do?

File: GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Controller/PurchaseController.cs [hunk 6]
```diff
@@ -2,6 +2,8 @@
 using System.Linq;
 using UnityEngine;
 using UnityEngine.Purchasing;
+using UnityEngine.Purchasing.Security;
+using IGooglePlayStoreExtensions = Google.Play.Billing.IGooglePlayStoreExtensions;
```
Comment (Qing451800) [inline GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Controller/PurchaseController.cs]: Do not import like this. You can fully specify the namespace.

File: GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Controller/PurchaseController.cs [hunk 7]
```diff
@@ -34,6 +37,11 @@ private void InitializePurchasing()
         // Create a builder, first passing in a suite of google play store.
         var builder = ConfigurationBuilder.Instance(Google.Play.Billing.GooglePlayStoreModule.Instance());
 
+        //Enable deferred purchases
+        builder.Configure<Google.Play.Billing.IGooglePlayConfiguration>()
+            .EnableDeferredPurchase();
+        Debug.Log("Deferred purchase enabled");
```
Comment (Qing451800) [inline GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Controller/PurchaseController.cs]: This log is not useful.

File: GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Controller/PurchaseController.cs [hunk 8]
```diff
@@ -117,73 +159,138 @@ public void OnInitializeFailed(InitializationFailureReason error)
...
+            // On Google Play, result has a single product ID.
+            var result = validator.Validate(receipt);
+            // For informational purposes, we list the receipt(s).
+            Debug.Log("Receipt is valid. Contents:");
+            foreach (IPurchaseReceipt productReceipt in result)
+            {
+                Debug.Log(productReceipt.productID);
+                Debug.Log(productReceipt.purchaseDate);
+                Debug.Log(productReceipt.transactionID);
+            }
```
Comment (Qing451800) [inline GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Controller/PurchaseController.cs:233]: return true;

you don't need validPurchase bool variable

File: GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Objects/GameData.cs [hunk 9]
```diff
@@ -112,8 +107,15 @@ public void IncreaseCoinsOwned(int increaseAmount)
...
     private void UpdateCoinText()
     {
-        Object.FindObjectOfType<GameManager>().SetCoins();
-        Object.FindObjectOfType<StoreController>().SetCoins();
+        Object.FindObjectOfType<GameManager>().SetCoinsBasedOnGameData();
+        Object.FindObjectOfType<StoreController>().SetCoinsBasedOnGameData();
+    }
+
+    // Purchase coins in store item.
+    public void PurchaseCoins(CoinList.Coin coinToPurchase)
```
Comment (Qing451800) [inline GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Objects/GameData.cs]: Rename this method. You're not purchasing here, i.e. UpdateCoins.

File: GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Controller/PurchaseController.cs [hunk 10]
```diff
@@ -117,73 +159,138 @@ public void OnInitializeFailed(InitializationFailureReason error)
...
+            gameData.PurchaseCar(car);
+            return;
+        }
+
+        // Check if a subscription has been purchased by this user.
+        foreach (var subscription in SubscriptionList.List.Where(subscription => string.Equals(productId,
+            subscription.ProductId,
+            StringComparison.Ordinal)))
+        {
+            gameData.SubscriptTo(subscription);
```
Comment (Qing451800) [inline GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Controller/PurchaseController.cs]: rename the method.

File: GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Objects/GameData.cs [hunk 11]
```diff
@@ -139,7 +138,7 @@ public bool CheckCarOwnership(CarList.Car car)
     }
 
     // Change car in use and apply the changes to the play page.
-    public void ChangeCar(CarList.Car targetCar)
+    public void ChangeCarInUse(CarList.Car targetCar)
```
Comment (Qing451800) [inline GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Objects/GameData.cs]: UpdateCarInuse

File: GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Objects/GameData.cs [hunk 12]
```diff
@@ -169,8 +168,8 @@ public void SetBackgroundBasedOnGameData()
         ChangeBackground(BackgroundList.List[(int) backgroundInUseName]);
     }
 
-    // Subscribe to a subscription.
-    public void SubscriptTo(SubscriptionList.Subscription targetSubscription)
+    // Update current subscription to target subscription.
```
Comment (Qing451800) [inline GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Objects/GameData.cs:168]: Use Update for all similar senarios

File: GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Objects/GameData.cs [hunk 13]
```diff
@@ -111,8 +111,8 @@ private void UpdateCoinText()
         Object.FindObjectOfType<StoreController>().SetCoinsBasedOnGameData();
     }
 
-    // Purchase coins in store item.
-    public void PurchaseCoins(CoinList.Coin coinToPurchase)
+    // Upgrade coins when player purchases coins.
+    public void UpgradeCoins(CoinList.Coin coinToPurchase)
```
Comment (Qing451800) [inline GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Objects/GameData.cs]: UpdateCoins

**Review by Qing451800 — APPROVED** [review]

File: GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Objects/GameData.cs (hunk 13 above)
Comment (zkzhao333) [reply GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Objects/GameData.cs]: Sorry, that is a typo.

**Pull request merged**